### DISCUSSION TITLE
Can anybody help me with a little more fleshed out pre consumption/post consumption scripting?

### DISCUSSION THREAD

**Discussion by zombiehoffa** [jonaswinkler/paperless-ng#361] [Support]
I am at a loss from the examples of how I would interact with paperless directly from those scripts. For example, I want to have multiple sub folders in my consumption directory and apply tags/correspondents depending on what directory the document was ingested from. I think I would do it in the post-consumption script but beyond that I don't really have an inkling as to where to go.

**Comment by andbez** [comment]
1. Correspondents are set automagically, when paperless will be trained by several documents.
2. Tags depending on Consuming-Dir can be set with a special Setting: See https://paperless-ng.readthedocs.io/en/latest/configuration.html#paths-and-folders / Parameter PAPERLESS_CONSUMER_SUBDIRS_AS_TAGS
3. I wrote a post-hook script in python to modify data in the paperless-database (but only for postgres, not sqlite). The Consuming-Dir is not available at post-hook-workflow, so that won't help you to set the correspondent.

**Reply by zombiehoffa** [reply]
Thanks, I'll take a look at that and play with it a bit. As long as the correspondents matching is high accuracy that should work for me. Honestly, I am not sure how I missed this looking through the docs, leaving me to think I had to do it in a processing script.

**Comment by jonaswinkler** [comment]
I think the current state of pre and post consumption scripts is pretty abysmal.

- They're hard to use with docker (you have to get them into the container)
- They don't have database access
- They can't influence the consumption process or the outcome thereof

A MUCH better way to do these pre/post scripts would be the following:

- Instead of a shell script, provide a python script that paperless imports (I'm not exactly sure if that's possible).
- This python script provides methods for various events in the consumption pipeline (pre/post/maybe more).
- This script would have access to all classes / methods that paperless has as well, including classes to access the database independent of the actual backend.
- This script could raise exceptions to cancel consumption.

You could do pretty much anything with these.

I'm not sure about security implications.

Does that sound like something we need?

**Reply by zombiehoffa** [reply]
That does sound way better (especially if it is well documented, so dabblers like myself can  figure out how to make it work). As far as whether it is something we need, I am honestly not sure yet because I have only been playing with this for a few hours and have given up on trying to make overly complicated mayan work so I don't have a good feel for what my work flow is actually going to end up looking like. Andbez's suggestion is probably going to get me through my current problem, especially if correspondent detection works well.

**Reply by andbez** [reply]
A more flexible way for pre- and post-hook-scripts would be nice and your suggestion sounds good. But i think the use-case for post-hook and pre-hook is limited. 
What can you change in paperless-Document-Data in post-hook? Just the Metadata depending from the information which is in the content and title of the document. And the Tag(s)/Correspondent/DocumentType should be set automagically. So only the ASN is not set automagically (Thats the reason, why i developed a script).
For Pre-Hook there have been some discussions regarding splitting of pdfs, set Metadata from Consume-Dir or delete empty pages, but that doesn't work without more dependencies. For pre-Hook i think it would be better to establish a workflow outside of paperless should be prefered for now (even when i think some basic pdf processes can be integrated into paperless-ng in the future) .

Just my 2 cents.
